subprojects/execution/src/test/groovy/org/gradle/internal/execution/steps/CaptureStateBeforeExecutionStepTest.groovy

```diff
@@ -0,0 +1,149 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.execution.steps
+
+import com.google.common.collect.ImmutableSortedMap
+import org.gradle.internal.classloader.ClassLoaderHierarchyHasher
+import org.gradle.internal.execution.AfterPreviousExecutionContext
+import org.gradle.internal.execution.BeforeExecutionContext
+import org.gradle.internal.execution.CachingResult
+import org.gradle.internal.execution.Step
+import org.gradle.internal.execution.UnitOfWork
+import org.gradle.internal.fingerprint.CurrentFileCollectionFingerprint
+import org.gradle.internal.fingerprint.impl.AbsolutePathFingerprintingStrategy
+import org.gradle.internal.hash.HashCode
+import org.gradle.internal.snapshot.FileSystemSnapshot
+import org.gradle.internal.snapshot.ValueSnapshot
+import org.gradle.internal.snapshot.ValueSnapshotter
+import org.gradle.internal.snapshot.impl.ImplementationSnapshot
+import spock.lang.Specification
+import spock.lang.Subject
+
+@Subject(CaptureStateBeforeExecutionStep)
+class CaptureStateBeforeExecutionStepTest extends Specification {
+
+    def classloaderHierarchyHasher = Mock(ClassLoaderHierarchyHasher)
+    def valueSnapshotter = Mock(ValueSnapshotter)
+    def context = Mock(AfterPreviousExecutionContext)
+    def work = Mock(UnitOfWork)
+    def result = Mock(CachingResult)
+    Step<BeforeExecutionContext, CachingResult> delegate = Mock()
+    def implementationSnapshot = ImplementationSnapshot.of("MyWorkClass", HashCode.fromInt(1234))
+
+    def step = new CaptureStateBeforeExecutionStep(classloaderHierarchyHasher, valueSnapshotter, delegate)
+
+    def "no state is captured when task history is not maintained"() {
+        when:
+        step.execute(context)
+        then:
+        1 * context.work >> work
+        1 * work.isTaskHistoryMaintained() >> false
+        1 * delegate.execute { BeforeExecutionContext context ->
+            context.beforeExecutionState.empty
+        } >> result
+        0 * _
+    }
+
+    def "implementations are snapshotted"() {
+        def additionalImplementations = [
+            ImplementationSnapshot.of("FirstAction", HashCode.fromInt(2345)),
+            ImplementationSnapshot.of("SecondAction", HashCode.fromInt(3456))
+        ]
+
+        when:
+        step.execute(context)
+        then:
+        1 * work.visitImplementations(_) >> { UnitOfWork.ImplementationVisitor visitor ->
+            visitor.visitImplementation(implementationSnapshot)
+            additionalImplementations.each {
+                visitor.visitAdditionalImplementation(it)
+            }
+        }
+        interaction { fingerprintInputs() }
+        1 * delegate.execute { BeforeExecutionContext context ->
```

Comment:
It is confusing that we have a `context` parameter here referring to the `BeforeExecutionContext` and a `context` field referring to the `AfterPreviousExecutionContext`. They should be named differently.

Reply:
I renamed both `context` variables.